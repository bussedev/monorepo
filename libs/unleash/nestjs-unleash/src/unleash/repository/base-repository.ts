// TODO: persist on disk

export abstract class BaseRepository<T extends { id: string }> {
  private items: T[] = []

  updateOrCreate(id: string, item: T): void {
    // eslint-disable-next-line unicorn/prefer-array-some
    this.find(item.id) ? this.update(id, item) : this.create(item)
  }

  create(item: T): void {
    this.items.push(item)
  }

  update(id: string, item: T): void {
    this.items[this.findIndex(id)] = item
  }

  delete(id: string): void {
    this.items.splice(this.findIndex(id), 1)
  }

  private findIndex(id: string) {
    return this.items.findIndex((item) => item.id === id)
  }

  find(id: string): T | undefined {
    return this.items.find((item) => item.id === id)
  }

  findAll(): T[] {
    return this.items
  }

  flushAll(): void {
    this.items = []
  }
}
