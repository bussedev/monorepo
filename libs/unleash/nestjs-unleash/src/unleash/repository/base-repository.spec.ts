import { BaseRepository } from './base-repository'

interface MyEntity {
  id: string
  name: string
}

class MyRepository extends BaseRepository<MyEntity> {}

const row: MyEntity = { id: '1', name: 'my name' }

describe('BaseRepository', () => {
  let repository: BaseRepository<MyEntity>

  beforeEach(() => {
    repository = new MyRepository()
  })

  test('updateOrCreate()', () => {
    repository.updateOrCreate('1', row)
    expect(repository).toMatchSnapshot()

    repository.updateOrCreate('1', { id: '1', name: 'my new name' })
    expect(repository).toMatchSnapshot()
  })

  test('create()', () => {
    repository.create(row)
    expect(repository).toMatchSnapshot()
  })

  test('update()', () => {
    repository.create(row)
    repository.update('1', { id: '1', name: 'updated' })
    expect(repository).toMatchSnapshot()
  })

  test('delete()', () => {
    repository.create(row)
    repository.delete('1')
    expect(repository.findAll()).toHaveLength(0)
  })

  test('find()', () => {
    expect(repository.find('1')).toBeUndefined()

    repository.create(row)
    expect(repository.find('1')).toEqual(row)
  })

  test('findAll()', () => {
    expect(repository.findAll()).toHaveLength(0)

    repository.create(row)
    repository.create({ id: '2', name: 'other' })
    expect(repository.findAll()).toMatchSnapshot()
  })
})
