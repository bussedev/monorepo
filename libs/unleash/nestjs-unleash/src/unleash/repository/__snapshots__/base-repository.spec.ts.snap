// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseRepository create() 1`] = `
MyRepository {
  "items": [
    {
      "id": "1",
      "name": "my name",
    },
  ],
}
`;

exports[`BaseRepository findAll() 1`] = `
[
  {
    "id": "1",
    "name": "my name",
  },
  {
    "id": "2",
    "name": "other",
  },
]
`;

exports[`BaseRepository update() 1`] = `
MyRepository {
  "items": [
    {
      "id": "1",
      "name": "updated",
    },
  ],
}
`;

exports[`BaseRepository updateOrCreate() 1`] = `
MyRepository {
  "items": [
    {
      "id": "1",
      "name": "my name",
    },
  ],
}
`;

exports[`BaseRepository updateOrCreate() 2`] = `
MyRepository {
  "items": [
    {
      "id": "1",
      "name": "my new name",
    },
  ],
}
`;
